File: generate-sitemap.py
#!/usr/bin/env python3
"""
FIXED Sitemap Generator for Taawidaty
ACTION: Generates ONLY valid routes that exist in the React app.
REMOVED: Medication database loop (caused 5000+ 404 errors).
Run with: python3 generate-sitemap.py
"""

from datetime import datetime
from pathlib import Path

BASE_URL = 'https://taawidaty.ma'

# The ONLY real routes in your application (verified from App.tsx)
VALID_ROUTES = [
    # Core Pages
    "/",
    "/prix-medicaments",
    "/blog",
    "/about-us",
    "/contact-us",
    
    # Blog Posts (All 8 posts)
    "/blog/guide-remboursement-cnss",
    "/blog/guide-remboursement-cnops",
    "/blog/difference-cnss-cnops",
    "/blog/comprendre-ppv-ppm-maroc",
    "/blog/medicament-generique-efficacite",
    "/blog/comprendre-ticket-moderateur",
    "/blog/medicaments-non-remboursables",
    "/blog/lire-ordonnance-maroc",
    
    # Author & Legal Pages
    "/author",
    "/privacy-policy",
    "/terms-of-service",
    "/medical-disclaimer",
    "/editorial-policy",
    "/cookie-preferences"
]

def generate_sitemap():
    """Generate clean sitemap.xml with only valid routes"""
    today = datetime.now().strftime('%Y-%m-%d')
    
    sitemap = '''<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:xhtml="http://www.w3.org/1999/xhtml">
'''
    
    for route in VALID_ROUTES:
        # Determine priority and frequency based on route type
        priority = "0.5"
        changefreq = "monthly"
        
        if route == "/":
            priority = "1.0"
            changefreq = "daily"
        elif route == "/prix-medicaments":
            priority = "0.9"
            changefreq = "weekly"
        elif route == "/blog":
            priority = "0.8"
            changefreq = "weekly"
        elif route.startswith("/blog/"):
            priority = "0.7"
            changefreq = "weekly"
        elif route == "/author":
            priority = "0.6"
            changefreq = "monthly"
        elif route in ["/about-us", "/contact-us"]:
            priority = "0.5"
            changefreq = "monthly"
        elif route in ["/privacy-policy", "/terms-of-service", "/medical-disclaimer", "/editorial-policy", "/cookie-preferences"]:
            priority = "0.3"
            changefreq = "yearly"

        # Build full URL
        full_url = f"{BASE_URL}{route}" if route != "/" else BASE_URL
        
        sitemap += f'''  <url>
    <loc>{full_url}</loc>
    <lastmod>{today}</lastmod>
    <changefreq>{changefreq}</changefreq>
    <priority>{priority}</priority>
    <xhtml:link rel="alternate" hreflang="fr" href="{full_url}"/>
    <xhtml:link rel="alternate" hreflang="ar" href="{full_url}"/>
    <xhtml:link rel="alternate" hreflang="x-default" href="{full_url}"/>
  </url>
'''
    
    sitemap += '</urlset>'
    
    # Write main sitemap
    Path('public/sitemap.xml').write_text(sitemap, encoding='utf-8')
    
    # Clean up old invalid sitemaps if they exist
    deleted_count = 0
    for i in range(1, 20):
        old_file = Path(f'public/sitemap-{i}.xml')
        if old_file.exists():
            old_file.unlink()
            deleted_count += 1
            print(f"🗑️  Deleted invalid file: {old_file}")

    print(f"\n✅ CLEAN Sitemap generated successfully!")
    print(f"📊 Total Valid URLs: {len(VALID_ROUTES)}")
    print(f"🚫 Fake medication URLs removed: ALL (5000+)")
    if deleted_count > 0:
        print(f"🗑️  Deleted {deleted_count} old sitemap files")
    print(f"📄 File size: ~{len(sitemap)} bytes")
    print(f"\n🔍 Next steps:")
    print(f"   1. Deploy this sitemap")
    print(f"   2. Submit to Google Search Console")
    print(f"   3. Request removal of /prix/ prefix in GSC")

if __name__ == '__main__':
    print("🚀 Generating FIXED sitemap for Taawidaty...\n")
    generate_sitemap()
if __name__ == '__main__':
    print("🚀 Generating FIXED sitemap for Taawidaty...\n")
    generate_sitemap()

File: generate-sitemap.mjs
/**
 * Sitemap Generator for Taawidaty
 * Generates sitemap.xml with all medication URLs for SEO
 * Also creates bing-urls.txt for Bing Webmaster Tools
 * Run with: node generate-sitemap.mjs
 */

import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const BASE_URL = 'https://taawidaty.ma'; // Production domain

function generateMedicationSlug(name) {
  return name
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[^a-z0-9\s-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .trim();
}

function generateSitemap() {
  const medications = [];
  
  try {
    // Load CNOPS medications
    const cnopsData = JSON.parse(readFileSync(join(__dirname, 'src/data/medications-cnops.json'), 'utf-8'));
    medications.push(...cnopsData);
  } catch (error) {
    console.warn('Could not load CNOPS medications:', error.message);
  }

  try {
    // Load CNSS medications (avoiding duplicates)
    const cnssData = JSON.parse(readFileSync(join(__dirname, 'src/data/medications-cnss.json'), 'utf-8'));
    const existingNames = new Set(medications.map(m => m.name));
    cnssData.forEach(med => {
      if (!existingNames.has(med.name)) {
        medications.push(med);
        existingNames.add(med.name);
      }
    });
  } catch (error) {
    console.warn('Could not load CNSS medications:', error.message);
  }

  const now = new Date().toISOString();

  // Generate XML sitemap
  let sitemap = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:xhtml="http://www.w3.org/1999/xhtml"
        xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">
  
  <!-- Main Pages -->
  <url>
    <loc>${BASE_URL}/</loc>
    <lastmod>${now}</lastmod>
    <changefreq>daily</changefreq>
    <priority>1.0</priority>
    <xhtml:link rel="alternate" hreflang="fr" href="${BASE_URL}/" />
    <xhtml:link rel="alternate" hreflang="ar" href="${BASE_URL}/" />
  </url>
  
  <url>
    <loc>${BASE_URL}/prix-medicaments</loc>
    <lastmod>${now}</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.9</priority>
    <xhtml:link rel="alternate" hreflang="fr" href="${BASE_URL}/prix-medicaments" />
    <xhtml:link rel="alternate" hreflang="ar" href="${BASE_URL}/prix-medicaments" />
  </url>
  
  <url>
    <loc>${BASE_URL}/faq-cnops</loc>
    <lastmod>${now}</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.7</priority>
  </url>
  
  <url>
    <loc>${BASE_URL}/faq-cnss</loc>
    <lastmod>${now}</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.7</priority>
  </url>
  
  <!-- Medication Pages (Top ${Math.min(medications.length, 5000)}) -->
`;

  // Add medication URLs (limit to 5000 for sitemap best practices)
  const topMedications = medications
    .sort((a, b) => b.ppv - a.ppv) // Sort by price (assuming higher price = more searched)
    .slice(0, 5000);

  topMedications.forEach(medication => {
    const slug = generateMedicationSlug(medication.name);
    const priority = medication.ppv > 100 ? '0.8' : '0.6'; // Higher priority for expensive meds
    
    sitemap += `  <url>
    <loc>${BASE_URL}/prix/${slug}</loc>
    <lastmod>${now}</lastmod>
    <changefreq>monthly</changefreq>
    <priority>${priority}</priority>
    <xhtml:link rel="alternate" hreflang="fr" href="${BASE_URL}/prix/${slug}" />
    <xhtml:link rel="alternate" hreflang="ar" href="${BASE_URL}/prix/${slug}" />
  </url>
`;
  });

  sitemap += `</urlset>`;

  // Write sitemap
  const sitemapPath = join(__dirname, 'public/sitemap.xml');
  writeFileSync(sitemapPath, sitemap);

  // Generate Bing URL list (plain text file)
  let bingUrls = `# Bing Webmaster Tools URL List
# Generated: ${new Date().toISOString()}
# Total URLs: ${topMedications.length + 4}

# Main Pages
${BASE_URL}/
${BASE_URL}/prix-medicaments
${BASE_URL}/faq-cnops
${BASE_URL}/faq-cnss

# Medication Pages (Top ${topMedications.length})
`;

  topMedications.forEach(medication => {
    const slug = generateMedicationSlug(medication.name);
    bingUrls += `${BASE_URL}/prix/${slug}\n`;
  });

  const bingUrlsPath = join(__dirname, 'public/bing-urls.txt');
  writeFileSync(bingUrlsPath, bingUrls);

  console.log(`✅ Sitemap generated successfully!`);
  console.log(`📊 Total URLs: ${topMedications.length + 4}`);
  console.log(`📝 Medications indexed: ${topMedications.length}`);
  console.log(`📄 Sitemap: ${sitemapPath}`);
  console.log(`📄 Bing URLs: ${bingUrlsPath}`);
  console.log(`\n🔗 Submit to:`);
  console.log(`   Google: https://search.google.com/search-console`);
  console.log(`   Bing: https://www.bing.com/webmasters`);
}

// Run the generator
generateSitemap();
